### DISCUSSION TITLE
Using ruff in pre-commit while supporting Python 3.6

### DISCUSSION THREAD

**Discussion by diazona** [bugen/pypipe#22] [General]
Hi @bugen, I thought this would be a good place to continue the discussion from https://github.com/bugen/pypipe/pull/20#issuecomment-1803914788 about how to use the ruff pre-commit hook while supporting Python 3.6.

Let me lay out all the information I know about this - some of it you might also know already, but hopefully the rest is useful - and you can decide how to proceed.

# ruff `target-version`

ruff needs to [target a specific Python version](https://docs.astral.sh/ruff/settings/#target-version). It won't recommend changes to the code that would be invalid in earlier Python versions than the target one. Current versions of ruff only support specifying Python 3.7 or later as the target version. However, in order to keep your code compatible with Python 3.6, you should set the target version to 3.6; otherwise there is a risk that ruff will insist you change something in a way that would be invalid in Python 3.6.

So, as long as you want to ensure that the code of pypipe always passes ruff's checks, you cannot safely use a recent version of ruff. You could, of course, continue to do what you're doing right now, namely setting `target-version` to Python 3.10 and just hoping that ruff does not require you to make any changes that would be incompatible with Python 3.6. Personally I wouldn't call that "safe", though.

If you want to continue with the "unsafe" approach, that's totally fine. But otherwise, you have to do one of the following:

- drop support for Python 3.6 (which I know you [said you don't want to do yet](https://github.com/bugen/pypipe/pull/20#issuecomment-1803914788))
- allow ruff to fail when run on the code
- use an earlier version of ruff that allows setting the target version to Python 3.6

# ruff in pre-commit

If you're going to use ruff as a pre-commit hook, the second option - allow ruff to fail - is not realistic, because the whole point of pre-commit is to (try to) ensure that the hooks pass on every commit. So you're left with using an earlier version of ruff and setting the `target-version` to Python 3.6 (or, really, don't set it at all, and allow it to use the target version specified in the `project.requires-python` key of `pyproject.toml`. From my experience with another project, I know that the latest version of ruff which lets you do that is 0.0.233. So your ruff pre-commit hook would need to look something like this (copied from [my other project](https://github.com/diazona/setuptools-pyproject-migration/blob/main/.pre-commit-config.yaml#L17-L26)):

```yaml
- repo: https://github.com/astral-sh/ruff-pre-commit
  # Last version of ruff which will accept "py36" as target-version
  rev: v0.0.233
  hooks:
  - id: ruff
    args:
    - "--fix"
    # When updating to ruff v0.0.244 or higher, this can be enabled to improve
    # change detection by pre-commit
    #- "--exit-non-zero-on-fix"
```

Then there is the question of how you actually run these pre-commit hooks. I mean, sure, ideally all developers _should_ be installing the hooks locally and running them locally. But that doesn't always happen. So it is quite useful for you to have a CI job that runs pre-commit.

## Using pre-commit.ci

The most common way to have a CI job that runs pre-commit is to register your project with [pre-commit.ci](https://pre-commit.ci). It's a service that will automatically run pre-commit for you on every pull request (or maybe on every push?) and on every update to the main branch. Pretty much all you have to do is register and approve that service to have access to your repo, and then everything will be set up for you - it's quite easy. You don't have to set up any workflows or runners or anything.

However, part of the pre-commit.ci service is that it will periodically make a pull request that updates all your pre-commit hooks to the latest versions. (Basically, this is because every hook used in every project that uses pre-commit.ci has to be cached by them, and they don't want to have to cache many very old hook versions.) You can [configure how often these pull requests happen](https://pre-commit.ci/#configuration-autoupdate_schedule), but you cannot disable them. So if you're using an older version of ruff, like you have to do to safely stay compatible with Python 3.6, the service will periodically make pull requests trying to update the ruff hook (and sometimes other hooks). You'll have to either manually modify those pull requests to "reset" the version of the ruff pre-commit hook back to 0.0.233, or reject them.

## Using Github Actions (or pre-commit.ci lite)

The alternative is to set up your own Github Actions job that runs pre-commit. You can use [pre-commit.ci lite](https://pre-commit.ci/lite.html) to help with this, or you can do it manually. Either way, you'll have to add a step to one of your workflow files, presumably `ci.yml`. I don't think it should be too hard, especially if you follow the "pre-commit.ci lite" instructions. However, this way you won't get automatic version updates to any pre-commit hooks at all.

**Comment by bugen** [comment]
@diazona Thank you for the detailed explanation. 

Here is my proposal:

- Currently, keep support for Python 3.6.
- However, execute ruff checks based on Python 3.10.
- Use pre-commit.ci for CI jobs.
- For pytest, run tests on both Python 3.6 and Python 3.10.

Before discussing ruff, it's important to address testing on Python 3.6. While I currently perform tests locally with Python 3.6, considering pull requests, it's crucial to automate tests on Python 3.6 in CI. I would like to modify the release.yml to include tests on Python 3.6, in addition to the existing tests on Python 3.10. Furthermore, when creating the pre-commit CI job, I would like to ensure that it includes testing on Python 3.6 as well.

Regarding ruff, I believe it's sufficient to perform checks based on Python 3.10. From a linter perspective, I don't anticipate significant differences between Python 3.6 and 3.10. Additionally, due to Python's backward compatibility, syntax considered invalid in Python 3.10 should also be invalid in Python 3.6. Therefore, I consider the checks in Python 3.10 to encompass those in Python 3.6. Considering the possibility of utilizing syntax or features introduced in Python 3.7 or later during pypipe modifications, in such cases, ruff checks may pass, but errors are expected in Python 3.6 tests (pytest) if the tests are properly written.

Out of curiosity, I checked pypipe with ruff version 0.0.233:
```
pypipe.py:216:89: E501 Line too long (94 > 88 characters)
pypipe.py:218:89: E501 Line too long (91 > 88 characters)
pypipe.py:222:89: E501 Line too long (89 > 88 characters)
pypipe.py:472:89: E501 Line too long (105 > 88 characters)
pypipe.py:476:89: E501 Line too long (103 > 88 characters)
pypipe.py:480:89: E501 Line too long (115 > 88 characters)
pypipe.py:572:89: E501 Line too long (89 > 88 characters)
pypipe.py:758:89: E501 Line too long (93 > 88 characters)
Found 8 error(s).
```
These errors don't impact functionality. While not detected in the latest version of ruff, I suspect this difference is due to variations in ruff's checking policies. Personally, I prefer these errors not to be detected, so I am inclined to use the newer version of ruff. If using the latest version of ruff, as you suggested, employing pre-commit.ci in the CI job seems like a better idea.

Please feel free to point out any inaccuracies in my understanding.

**Reply by diazona** [reply]
> Before discussing ruff, it's important to address testing on Python 3.6. While I currently perform tests locally with Python 3.6, considering pull requests, it's crucial to automate tests on Python 3.6 in CI. I would like to modify the release.yml to include tests on Python 3.6, in addition to the existing tests on Python 3.10. Furthermore, when creating the pre-commit CI job, I would like to ensure that it includes testing on Python 3.6 as well.

Adding tests on Python 3.6 to CI will be no problem :+1: However, I think it could (and maybe should) be considered a separate issue from what I'm discussing here, because as a general guideline, any checks which need to be performed on multiple versions of Python are probably not suitable as pre-commit hooks. There are two reasons:

- First, those checks are often slow. pre-commit hooks have to run every single time a developer makes a commit, and if the hooks take more than a second or so, maybe two seconds at the most, it can get annoying having to wait for them. (In particular, running pytest is almost certainly too slow to be a good pre-commit hook.)
- Also, many of the checks that run in pre-commit can actually modify the code when they run - for example, many linting errors can be automatically fixed by the linter hook. However, if you run multiple instances of the same hook that target different Python versions, those instances might fix the code in different ways, so they would clash with each other and make it impossible to commit anything.

Given those reasons, I would suggest that any multi-version checks, such as running pytest, should be added _only_ to the CI jobs, namely `release.yml` and `ci.yml`. (And of course developers can also run them locally, just not on every commit.) For pre-commit hooks, it's better to limit yourself to checks which run quickly and only need to be run for a single version of Python.

> From a linter perspective, I don't anticipate significant differences between Python 3.6 and 3.10. Additionally, due to Python's backward compatibility, syntax considered invalid in Python 3.10 should also be invalid in Python 3.6. Therefore, I consider the checks in Python 3.10 to encompass those in Python 3.6.

It's true that syntax which is invalid in Python 3.10 would also be invalid in 3.6, and if all ruff did was check for invalid syntax, then yes, you'd have a point. However the argument I was making is based on the assumption that ruff is not being used only to check for things that are invalid, it can also check for valid but non-recommended code, and it can also be used to _fix_ the code. And the problem there is that, if you give ruff a target version of Python 3.10, it might complain that something which is technically valid should be changed because there is a better alternative available in Python 3.10, and/or it might fix the code in a way that is valid in Python 3.10 but not in 3.6.

I don't know of a real example offhand, but to invent a fake example that shows the kind of thing I'm talking about: suppose you have this code snippet:

```python
from typing import List, Optional

class Animal:
    ...

def fill_hovercraft(animals: Optional[List[Animal]]) -> None:
    ...
```

With some hypothetical linter `fake-lint`, you could run `fake-lint --target-version py36` on that code and it would pass. But if you run `fake-lint --target-version py310` on that same code it would fail, because the recommended way of expressing optional types and parametrized container types changed (in Python 3.9). The linter would expect you to write this instead:

```python
class Animal:
    ...

def fill_hovercraft(animals: list[Animal] | None) -> None:
    ...
```

And if `fake-lint` is added as a pre-commit hook, then here you have a situation where you are actually _forbidden_ from committing code that is valid in Python 3.6, and _required_ to change it in a way that makes it invalid. That's the sort of thing I'm talking about.

Now, I did some tests and found that ruff does not make that _specific_ change. But there may be other changes it makes that are broadly similar: it would refuse to allow you to commit code that is valid in Python 3.6 unless you change it in a way that makes it invalid. That's why I'm saying that using Python 3.10 as the target version, even though you want to support Python 3.6, is "unsafe".

> ```
> pypipe.py:216:89: E501 Line too long (94 > 88 characters)
> ...
> ```
>
> While not detected in the latest version of ruff, I suspect this difference is due to variations in ruff's checking policies. Personally, I prefer these errors not to be detected, so I am inclined to use the newer version of ruff.

That's not really related to which version of ruff you use, it's related to how you configure it. The default configuration changed in v0.1 of ruff to stop checking line lengths by default (see https://github.com/astral-sh/ruff/pull/7900), but you can get the same behavior with older versions of ruff using the `--ignore E501` option or its equivalent in a [configuration file](https://docs.astral.sh/ruff/configuration/). Alternatively, you could set a custom maximum line length using either a command line option, like `--line-length 120` (or any value you like), or the equivalent in the config file.

---

Anyway, I guess my main point would be that, since you want to keep compatibility with Python 3.6, I think using ruff 0.0.233 is the choice that is most consistent with that goal. But it's your choice, and you can use the latest version of ruff if you want to, you should just be prepared for the possibility that you'll have to choose between that and Python 3.6 compatibility at some point, and it may come as a surprise.

**Comment by bugen** [comment]
@diazona  Thanks for your advice. I understand the issues you pointed out.
The upcoming release, v0.5.0, will be the final version of pypipe to support Python 3.6. Until then, I would like to keep this matter on hold. For versions of pypipe released after v0.5.0, I am considering supporting Python 3.8 or later. At that time, I would like to revisit the discussion regarding ruff and pre-commit.
